### PR TITLE
TypeChecker error when encoding functions with call options; tests

### LINKED ISSUE #8590: [solc] Internal compiler error when encountering non-encodable `f{gas: 2}` type
## Description
The following contract should issue an error about non-encodable type. Currently there is an internal compiler error.
```
contract C {
    function f() public {
		abi.encode(this.f{gas: 2});
    }
}
```

> Internal compiler error during compilation:
> /solidity/libsolidity/codegen/ContractCompiler.cpp(77): Throw in function void {anonymous}::StackHeightChecker::check()
> Dynamic exception type: boost::wrapexcept<solidity::langutil::InternalCompilerError>
> std::exception::what: I sense a disturbance in the stack: 2 vs 1
> [solidity::util::tag_comment*] = I sense a disturbance in the stack: 2 vs 1
> [harikrishnan@localhost]~/area51/solidity/Debug/solc% ./solc /tmp/encode.sol

The following contract (based on test/libsolidity/syntaxTests/viewPureChecker/gas_value_without_call.sol)

```
contract C {
    function f() public pure {
        bytes32 h = keccak256(abi.encodePacked(keccak256, f, this.f{gas: 2}, blockhash));
        h;
    }
}
```

should issue an error about not `this.f{gas: 2}` not being encodable. However, that is skipped.

> Error: This type cannot be encoded.
>   |
>   |         bytes32 h = keccak256(abi.encodePacked(keccak256, f, this.f{gas: 2}, blockhash));
>   |                                                ^^^^^^^^^
> 
> Error: This type cannot be encoded.
>   |
>   |         bytes32 h = keccak256(abi.encodePacked(keccak256, f, this.f{gas: 2}, blockhash));
>   |                                                           ^
> 
> Error: This type cannot be encoded.
>   |
>   |       
>   | bytes32 h = keccak256(abi.encodePacked(keccak256, f, this.f{gas: 2}, blockhash));

## Environment

- Compiler version: current develop.
## Steps to Reproduce

`solc input.sol`

## Notes

The problem also exists for the dot syntax for both value and error.

### PR DESCRIPTION
**Pull request by hrkrshnn**
Fixes https://github.com/ethereum/solidity/issues/8590

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -1884,6 +1884,23 @@ void TypeChecker::typeCheckABIEncodeFunctions(
 			}
 		}
 
+		// The case of function call options, e.g., f{gas: 2}.
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: Why did you not change `encodingType` instead?

Comment (ekpyron) [reply libsolidity/analysis/TypeChecker.cpp]: Because I told him not to :-). Change to what exactly :-)?

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp]: Ah :) Change to return nullptr if gas or value is set?

Comment (ekpyron) [reply libsolidity/analysis/TypeChecker.cpp]: Yeah, ok, it came up whether to set the kind of the ``encodingType`` back to ``SetGas`` and I thought that was a bad idea and suggested this here. But you're right, it'd be better to just change ``encodingType`` to actually return ``nullptr`` for these types.

Comment (hrkrshnn) [reply libsolidity/analysis/TypeChecker.cpp]: Ok, I will change FunctionType::encodingType() to return nullptr if it has call options.

**Review by chriseth — APPROVED** [review]

**Pull request merged**